src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -8498,19 +8487,22 @@ protected override void VisitForEachExpression(BoundForEachStatement node)
 
             MethodSymbol? reinferredGetEnumeratorMethod = null;
 
-            if (node.EnumeratorInfoOpt?.GetEnumeratorMethod is { IsExtensionMethod: true, Parameters: var parameters } enumeratorMethod)
+            if (node.EnumeratorInfoOpt?.GetEnumeratorInfo is { Method: { IsExtensionMethod: true, Parameters: var parameters } } enumeratorMethodInfo)
             {
                 // this is case 7
+                // We do not need to do this same analysis for non-extension methods because they do not have generic parameters that
+                // can be inferred from usage like extension methods can. We don't warn about default arguments at the call site, so
+                // there's nothing that can be learned from the non-extension case.
                 var (method, results, _) = VisitArguments(
                     node,
-                    ImmutableArray.Create(node.Expression),
+                    enumeratorMethodInfo.Arguments,
                     refKindsOpt: default,
                     parameters,
-                    argsToParamsOpt: default,
-                    defaultArguments: default,
+                    argsToParamsOpt: enumeratorMethodInfo.ArgsToParamsOpt,
+                    defaultArguments: enumeratorMethodInfo.DefaultArguments,
```

Comment:
Do we need a similar change for default arguments in disposal in using-declaration?

Reply:
I don't believe so, because I don't believe there is any way to observe a nullability issue there. We made the decision to not warn at the call site for default parameters that don't fit the type, and there is no way to get generic substitution at the method level because we don't allow extension methods for using statements/declarations. If there's a case you can think of I'm open to it, but we don't even analyze the method parameters at all today for those cases.

---
In reply to: [549493697](https://github.com/dotnet/roslyn/pull/50105#discussion_r549493697) [](ancestors = 549493697)